### PR TITLE
Enable capturing another process's stdio

### PR DESCRIPTION
**Pull request by joeyates**
* Adds an optional PID parameter to capture_io for standard_io capture.

### CODE REVIEW

**Comment by fishcakez** [comment]
@joeyates this is an interesting idea but unfortunately the implementation has a few issues that we may not be able to resolve.

The change does not guarantee to revert the group leader because the process calling `capture_io` could exit due to an exit signal and never call the `after` clause.

Also I am unsure about some race conditions when capturing another process's `:stdio`. It would be possible for multiple process to have interleaving `capture_io` calls on the same process. This could cause the group leader to be reset to an incorrect process. This isn't possible when capturing the `:stdio` of the calling process because the `capture_io` calls can only be nested and not interleaving.

It might be possible to handle these issues by using a lock in the CaptureServer so that each process could only be captured once. Technically this would break nested `capture_io` calls but I am unsure if that is an issue because it seems a strange thing to do.

**Comment by joeyates** [comment]
@fishcakez Thanks for your comments.

I agree that there is no way to provide guarantees about setting and restoring another process's group leader; nothing stops a developer using both `capture_io` and another 'manual' method for setting another process's group leader - at which point all bets are off.

I do, though, think that it would be useful to be able to capture a child processes' `:stdio`, so maybe the solution is to warn about the lack of guarantees and possible race conditions in the docs:

I.e.

    It is possible to capture another process's `:stdio`, but doing so requires caution:
    1) no checks are performed about multiple interleaved or overlapping calls to `capture_io`,
    2) if the process that calls `capture_io` dies, no cleanup will be performed.

**Comment by fishcakez** [comment]
I meant that interleaving `capture_io` calls are possible with this PR where two (or more) processes try to capture the io of the same process. This could easily occur if async tests try to capture the io of a process in a running application. We would need to provide the guarantee those 2 caveats are not possible to include this, i.e. test the comments I made above can not happen.

It is worth noting to anyone wanting something like this feature that process spawned inside a `capture_io` will have their IO captured because spawned process inherit the group leader of their parent at the time they are spawned. For example any IO generated by the child server below will be captured:
```elixir
capture_io(fn() ->
  {:ok, pid} = GenServer.start_link(..)
  GenServer.call(pid, ...)
  GenServer.stop(pid)
end)
```

**Comment by joeyates** [comment]
@fishcakez I believe that processes launched by a test itself do not suffer from the race conditions you mention - an example of this is the tests I added to this PR.

So, the problem would arise for processes that run for the duration of the test suite.

As `capture_io` for `:stderr` and other devices already carries the caveat that tests must be run with `async: false`, I suggest making the same proviso when capturing processes whose lifecycle is not controlled by the test.

**Comment by fishcakez** [comment]
@joeyates `:stderr` and other device having a mechanism to ensure they can only be run with `async: false` and provide a good error message when this is violated. We would need the same for this feature.

**Comment by lexmag** [comment]
I personally not sure if we should add this feature, I'm trying to think about it in terms of maintenance cost and interface. With this patch API feels quite overloaded and less intuitive.
My another concern is that it is possible to capture another process's stdio today, like in the example @fishcakez has shown before, or alternatively with:
```elixir
pid = ...
{:group_leader, group_leader} = Process.info(pid, :group_leader)
capture_io(fn() ->
try
  Process.group_leader(pid, Process.group_leader)
  ...
  Process.group_leader(pid, group_leader)
end)
```

File: lib/ex_unit/lib/ex_unit/capture_io.ex [hunk 1]
```diff
@@ -141,6 +154,29 @@ defmodule ExUnit.CaptureIO do
...
+    original_gl = Process.group_leader()
+    {:ok, capture_gl} = StringIO.open(input, capture_prompt: prompt_config)
+    case ExUnit.CaptureServer.process_capture_on(pid) do
+      {:ok, ref} ->
+        try do
+          Process.group_leader(pid, capture_gl)
+          do_capture_io(capture_gl, fun)
+        after
+          if Process.alive?(pid) do
+            Process.group_leader(pid, original_gl)
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/capture_io.ex]: Seems we're setting incorrect group leader here, leader of the called but not of the `pid`. 🤔

**Comment by fishcakez** [comment]
@lexmag are you saying the behaviour is confusing because an atom device name replaces the device for processes sending to that device by name but using a pid forwards `:stdio` from that pid? I think that is quite confusing. We would need a different API, perhaps `{:stdio | :standard_io, pid}` as argument. What is the maintenance concern?

**Comment by joeyates** [comment]
@fishcakez I think that @lexmag is (correctly) saying that the code cannot get the process's actual `group_leader` - it assumes that the process was launched by the testing process and so has the same group_leader.

While I have been able to add code to handle the race condition problems and the early exit of the calling process, I can't see a way of fixing this problem.

Currently, there seems to be no way in Elixir or Erlang to query a process's group leader. If that is in fact the case, I think I we hope to handle the case where one process wishes to capture another process's IO.

For this reason, I think it may be best to close this PR.

I've created a mini-library `capture_child_io` which handles the core use case:  capturing the io of a process created by the test. I'll add all the necessary caveats to the library's docs.

**Comment by josevalim** [comment]
```elixir
iex> Process.info(self(), :group_leader)
{:group_leader, #PID<0.26.0>}
iex> Process.info(Process.whereis(:user), :group_leader)
{:group_leader, #PID<0.9.0>}
```

Just be careful what happens if the process is dead.

**Comment by lexmag** [comment]
@fishcakez exactly, when we pass device we capture device, but with PID we capture its `:stdio`, it is possible to assume that user should pass PID of device.

> What is the maintenance concern?

We would have extra amount of code which is not an easy one (I assume this patch has race condition around `if Process.alive?` and `Process.group_leader` calls).
Though it already is possible to capture another process's stdio today:
```elixir
capture_io(fn() ->
  {:ok, pid} = GenServer.start_link(..)
  GenServer.call(pid, ...)
  GenServer.stop(pid)
end)
```
and
```elixir
pid = ...
{:group_leader, group_leader} = Process.info(pid, :group_leader)
capture_io(fn() ->
try
  Process.group_leader(pid, Process.group_leader)
  ...
  Process.group_leader(pid, group_leader)
end)
```
_@joeyates please see `Process.info/2` usage in the example above._ :bowtie:

Perhaps we should improve documentation about that.

**Comment by fishcakez** [comment]
I am not keen on the alternative example that calls `Process.group_leader` because it has the same issues as the initial commit in this PR (assuming the reset call is in an after clause). I don't think we should encourage that pattern.

Without the suggested feature here I think a user must link from the process calling `capture_io` in such a way that an exit in the calling process stops the captured process to provide good isolation and repeatability in the tests. If captured process is trapping exits this likely requires the calling process to be a parent or ancestor of the captured process.

With this feature, or the changing group leader alternative example, there is not good isolation because any process being spawned by the captured process will inherit the capture io group leader. However when the captured process's group leader is reset the spawned processes wont get their group leader reset. When capturing a named deviced or `:stdio` this is not an issue because either the group leader is not changed or all descendants will exit when the test/calling process does, so any processes that are on the capture io group leader will be cleaned up once the test exits. This issue really puts me off including this feature because it may make users tests fragile.

**Comment by joeyates** [comment]
I've updated the PR to add handling for correctly resetting the captured process's group leader.

I think there is one outstanding problem: as @fishcakez says, "any process being spawned by the captured process will inherit the capture io group leader", and the same is true for the alternative testing strategy:

```elixir
capture_io(fn() ->
  {:ok, pid} = GenServer.start_link(..)
  GenServer.call(pid, ...)
  GenServer.stop(pid)
end)
```

I can't see a solution for this problem.

**Comment by joeyates** [comment]
I've added a solution, but it's not pretty - it iterates over all processes on the node looking for errant spawned processes that need to have their group leader corrected.

**Comment by fishcakez** [comment]
Unfortunately trying to do that still has race conditions because another test might try to capture the io of a process we reset :(.

**Comment by fishcakez** [comment]
I am not sure we can make this solid enough, thoughts?

**Comment by fishcakez** [comment]
I am closing because this is too brittle :(.

**Pull request closed without merging**